x/posts/keeper/msg_server.go

```diff
@@ -0,0 +1,478 @@
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/desmos-labs/desmos/v3/x/posts/types"
+	subspacestypes "github.com/desmos-labs/desmos/v3/x/subspaces/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the stored MsgServer interface
+// for the provided k
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = &msgServer{}
+
+// CreatePost defines the rpc method for Msg/CreatePost
+func (k msgServer) CreatePost(goCtx context.Context, msg *types.MsgCreatePost) (*types.MsgCreatePostResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check the if the subspace exists
+	if !k.HasSubspace(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	author, err := sdk.AccAddressFromBech32(msg.Author)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid author address: %s", msg.Author)
+	}
+
+	// Check the permission to create content
+	if !k.HasPermission(ctx, msg.SubspaceID, author, subspacestypes.PermissionWrite) {
+		return nil, sdkerrors.Wrap(subspacestypes.ErrPermissionDenied, "you cannot create content inside this subspace")
+	}
+
+	// Get the next post id
+	postID, err := k.GetNextPostID(ctx, msg.SubspaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create and validate the post
+	post := types.NewPost(
+		msg.SubspaceID,
+		postID,
+		msg.ExternalID,
+		msg.Text,
+		msg.Author,
+		msg.ConversationID,
+		msg.Entities,
+		msg.ReferencedPosts,
+		msg.ReplySettings,
+		ctx.BlockTime(),
+		nil,
+	)
+	err = k.ValidatePost(ctx, post)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the post
+	k.SavePost(ctx, post)
+
+	// Update the id for the next post
+	k.SetNextPostID(ctx, msg.SubspaceID, post.ID+1)
+
+	// Unpack the attachments
+	attachments, err := types.UnpackAttachments(k.cdc, msg.Attachments)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid attachments: %s", err)
+	}
+
+	// Store the attachments
+	for _, content := range attachments {
+		_, err = k.storePostAttachment(ctx, post.SubspaceID, post.ID, content)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, sdk.MsgTypeURL(msg)),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Author),
+		),
+		sdk.NewEvent(
+			types.EventTypeCreatePost,
+			sdk.NewAttribute(types.AttributeKeySubspaceID, fmt.Sprintf("%d", msg.SubspaceID)),
+			sdk.NewAttribute(types.AttributeKeyPostID, fmt.Sprintf("%d", post.ID)),
+			sdk.NewAttribute(types.AttributeKeyAuthor, msg.Author),
+			sdk.NewAttribute(types.AttributeKeyCreationTime, post.CreationDate.Format(time.RFC3339)),
```

Comment:
I have a dump question, why timestamp should be indexed?

Reply:
This is done in order to allow clients to get the creation time from the returned events. Since it is based on the block time, and this is not know to the client, the only way we have to return it to clients (so that they can update the post information locally) is though block events that can later be parsed